router/core/graphql_prehandler.go

```diff
@@ -113,168 +85,36 @@ func (h *PreHandler) Handler(next http.Handler) http.Handler {
-			w.WriteHeader(statusCode)
-			writeRequestErrors(graphql.RequestErrorsFromError(errors.New("operation name is required when multiple operations are defined")), w, requestLogger)
-			return
-		}
-
-		// Add client info to trace span
-		clientName := ctrace.GetClientInfo(r.Header, "graphql-client-name", "apollographql-client-name", "unknown")
-		clientVersion := ctrace.GetClientInfo(r.Header, "graphql-client-version", "apollographql-client-version", "missing")
-
-		// Add client info to trace span attributes
-		span.SetAttributes(otel.WgClientName.String(clientName))
-		span.SetAttributes(otel.WgClientVersion.String(clientVersion))
-
-		// Add client info to metrics base fields
-		metricBaseFields = append(metricBaseFields, otel.WgClientName.String(clientName))
-		metricBaseFields = append(metricBaseFields, otel.WgClientVersion.String(clientVersion))
-
-		requestOperationNameBytes := unsafebytes.StringToBytes(requestOperationName)
-
-		normalizer := astnormalization.NewNormalizer(true, false)
-
-		if requestOperationName == "" {
-			normalizer.NormalizeOperation(doc, h.executor.Definition, report)
-		} else {
-			normalizer.NormalizeNamedOperation(doc, h.executor.Definition, requestOperationNameBytes, report)
-		}
-
-		if report.HasErrors() {
-			statusCode = http.StatusBadRequest
-			logInternalErrors(report, requestLogger)
-			w.WriteHeader(statusCode)
-			writeRequestErrorsFromReport(report, w, requestLogger)
-			return
-		}
-
-		hash := xxhash.New()
-
-		// add the operation name to the hash
-		// this is important for multi operation documents to have a different hash for each operation
-		// otherwise, the prepared plan cache would return the same plan for all operations
-		_, err = hash.Write(requestOperationNameBytes)
+		operation, err := h.parser.Parse(buf.Bytes())
 		if err != nil {
-			requestLogger.Error("hash write failed", zap.Error(err))
-			statusCode = http.StatusInternalServerError
-			w.WriteHeader(statusCode)
-			writeRequestErrors(graphql.RequestErrorsFromError(internalServerErrorErr), w, requestLogger)
-			return
-		}
-
-		printer := &astprinter.Printer{}
-
-		// create a hash of the query to use as a key for the prepared plan cache
-		// in this hash, we include the printed operation
-		// and the extracted variables (see below)
-		err = printer.Print(doc, h.executor.Definition, hash)
-		if err != nil {
-			requestLogger.Error("unable to print document", zap.Error(err))
-			statusCode = http.StatusInternalServerError
-			w.WriteHeader(statusCode)
-			writeRequestErrors(graphql.RequestErrorsFromError(internalServerErrorErr), w, requestLogger)
+			var reportErr ReportError
+			var inputErr *inputError
+			switch {
+			case errors.As(err, &inputErr):
+				statusCode = http.StatusBadRequest
+				requestLogger.Error(inputErr.Error())
+				w.WriteHeader(statusCode)
+				w.Write([]byte(inputErr.Error()))
+			case errors.As(err, &reportErr):
+				report := reportErr.Report()
```

Comment:
The report can contain internal and external errors.
According to the graphql-over-http spec, external errors should be propagated as 200 OK, whereas internal errors should be propagated as 500.